### PR TITLE
Implemented scopes for projection expressions

### PR DESCRIPTION
**Pull request by blackstardlb**

### CODE REVIEW

File: src/expressions/projectionExpressionChecker.js [hunk 1]
```diff
@@ -1,39 +1,135 @@
...
-    const core = this._getFactChecker().getAbundanceService().getCoreAPI()
-    const lawregContext = context
+    const scope = fact.scope ? fact.scope : 'single'
+    return this.scopeCheckers[scope].checkSubExpression(fact, ssid, context)
+  }
+}
 
+class BaseScopeProjectionExpressionChecker extends BaseSubExpressionChecker {
+  _checkProjectionIsValid (fact) {
+    // TODO current solution only requires last entry in context array so maybe this should just be a single property
```
Comment (pimotte) [inline src/expressions/projectionExpressionChecker.js:26]: Was this always the case? The reason to have context be an array is to handle nested created facts, right?

**Review by pimotte — COMMENTED** [review]
LGTM, one small question before we merge:)

File: src/expressions/projectionExpressionChecker.js (hunk 1 above)
Comment (blackstardlb) [reply src/expressions/projectionExpressionChecker.js:26]: I don't know why we ever needed that. this._getFactChecker().getCreatingActs(fact, ssid, context) can get all acts where fact was created. And the 'should project a series of CREATE facts' still passes.

Comment (blackstardlb) [reply src/expressions/projectionExpressionChecker.js:26]: Unless i'm missing something i don't see why we need it

Comment (pimotte) [reply src/expressions/projectionExpressionChecker.js:26]: Okay, so I don't know if it's necessary, but it enables stuff like (pseudo-model)
```
[car accident]: CREATE('[causing car]', ['victims car]'), 
[causing car]: CREATE([driver], [license plate], [passengers]), 
[victims car]: CREATE([driver], [license plate], [passengers]), 
[driver]: CREATE([license])
[passenger]: CREATE([license])
```

Then if you want to get the license of the driver that was in the causing car, you can do a projection with context `["[causing car]", "[driver]"]`, does that make sense?

Comment (blackstardlb) [reply src/expressions/projectionExpressionChecker.js:26]: I see i added a test for this and it does indeed fail good looking out :+1:

File: src/expressions/baseSubExpressionChecker.js [hunk 2]
```diff
@@ -36,4 +39,13 @@ export class BaseSubExpressionChecker {
   _getFactChecker () {
     return this.serviceProvider.factChecker
   }
+
+  /**
+   * Get fact checker
```
Comment (pimotte) [inline src/expressions/baseSubExpressionChecker.js:44]: Small copy/paste error?

File: src/expressions/projectionExpressionChecker.js [hunk 3]
```diff
@@ -1,39 +1,270 @@
...
-    const initialLink = await this._getFactChecker().checkFact(fact.context[0], ssid, lawregContext)
-    const caseLink = await fact.context.slice(1).reduce(async (previousCaseLink, currentContextFact) => {
-      if (previousCaseLink) {
-        const factContext = await core.get(previousCaseLink, ssid)
+  /**
+   * Get filtered creating acts
+   * @param {object} fact - the create expression
+   * @param {ssid} ssid - Identifies the actor
+   * @param {Context} context - Context of the action
+   * @return {Promise<Object<string, Object<string, *>>>} - Object where key is the act link and value is the provided facts
```
Comment (pimotte) [inline src/expressions/projectionExpressionChecker.js:46]: The we use this as an object anywhere? Maybe it makes sense for this to be an array of CreatingAct?

File: test/testUtils.js [hunk 4]
```diff
@@ -11,13 +11,15 @@ import { expect } from 'chai'
  * @param {string} actor - actor name
  * @param {string} act - description of the act to be taken
  * @param {function(string) : *} factResolver - Function used to resolve facts to fall back on if no other method is available. Defaults to always false
+ * @param {Object<string, number[]|number>} cases - Object where the keys are createable facts and the values are the action index to use for that fact
  * @returns {Step}
  */
-function takeAction (actor, act, factResolver = () => false) {
+function takeAction (actor, act, factResolver = () => false, cases = {}) {
```
Comment (pimotte) [inline test/testUtils.js:17]: Is there a reason you added the cases parameter? I think the behaviour could be obtained by a fact resolver, in fact, we even wrap it in this function. Maybe instead we could make a factory for factresolvers or something that handles this kind of thing?

**Review by pimotte — APPROVED** [review]
Some comments here and there, I would be okay with pushing any of them to a later PR as long as we do an issue or TODO for them.

**Review by pimotte — APPROVED** [review]

**Pull request merged**